File: JULY-2025/06-07-2025/p1865.go
type FindSumPairs struct {
  nums1, nums2 []int
  count map[int]int
}


func Constructor(nums1 []int, nums2 []int) FindSumPairs {
  count := make(map[int]int)
  for _, num := range nums2 {
    count[num]++
  }

  return FindSumPairs{
    nums1: nums1,
    nums2: nums2,
    count: count,
  }
}


func (f *FindSumPairs) Add(index int, val int)  {
  prevVal := f.nums2[index]
  f.count[prevVal]--

  newVal := prevVal+val
  f.count[newVal]++
  f.nums2[index] = newVal
}


func (f *FindSumPairs) Count(tot int) int {
  total := 0
  for _, num := range f.nums1 {
    if num > tot {
      continue
    }

    diff := tot-num
    if amount, ok := f.count[diff]; ok && amount > 0 {
      total+=amount
    }
  }

  return total
}


/**
 * Your FindSumPairs object will be instantiated and called as such:
 * obj := Constructor(nums1, nums2);
 * obj.Add(index,val);
 * param_2 := obj.Count(tot);
 */

File: JULY-2025/06-07-2025/p1865.kt
class FindSumPairs(nums1: IntArray, val nums2: IntArray) {
    
    //create a frequency map for nums1
    private val freq1: Map<Number, OccurrenceCount> = nums1.toList().groupingBy { it }.eachCount()

    fun add(index: Int, value: Int) {
        nums2[index] += value
    }

    //search for the complement of num in nums2. 
    fun count(tot: Int): Int {
        var count = 0
        for (num in nums2) {
            if (freq1.contains(tot - num)) {
                count += freq1[tot - num]!!
            }
        }
        return count
    }
}

typealias Number = Int
typealias OccurrenceCount = Int


/**
 * Your FindSumPairs object will be instantiated and called as such:
 * var obj = FindSumPairs(nums1, nums2)
 * obj.add(index,`val`)
 * var param_2 = obj.count(tot)
 */